src/lsdb/core/crossmatch/kdtree_match.py

```diff
@@ -0,0 +1,177 @@
+import numpy as np
+import pandas as pd
+import healpy as hp
+from sklearn.neighbors import KDTree
+
+from lsdb.core.crossmatch.abstract_crossmatch_algorithm import (
+    AbstractCrossmatchAlgorithm,
+)
+
+
+class KdTreeCrossmatch(AbstractCrossmatchAlgorithm):
+    """Nearest neighbor crossmatch using a K-D Tree"""
+
+    def crossmatch(
+        self,
+        n_neighbors: int = 1,
+        d_thresh: float = 0.01,
+    ) -> pd.DataFrame:
+        """Perform a cross-match between the data from two HEALPix pixels
+
+        Finds the n closest neighbors in the right catalog for each point in the left catalog that
+        are within a threshold distance by using a K-D Tree.
+
+        Args:
+            n_neighbors (int): The number of neighbors to find within each point
+            d_thresh (float): The threshold distance beyond which neighbors are not added
+
+        Returns:
+            A DataFrame from the left and right tables merged with one row for each pair of
+            neighbors found from cross-matching. The resulting table contains the columns from the
+            left table with the first suffix appended, the right columns with the second suffix, and
+            a `_DIST` column with the great circle separation between the points.
+        """
+
+        # get matching indices for cross-matched rows
+        left_idx, right_idx = self._find_crossmatch_indices(n_neighbors)
+
+        # filter indexes to only include rows with points within the distance threshold
+        (
+            distances,
+            left_ids_filtered,
+            right_ids_filtered,
+        ) = self._filter_indexes_to_threshold(left_idx, right_idx, d_thresh)
+
+        # rename columns so no same names during merging
+        self._rename_columns_with_suffix(self.left, self.suffixes[0])
+        self._rename_columns_with_suffix(self.right, self.suffixes[1])
+
+        # concat dataframes together
+        self.left.index.name = "_hipscat_index"
+        left_join_part = self.left.iloc[left_ids_filtered].reset_index()
+        right_join_part = self.right.iloc[right_ids_filtered].reset_index(drop=True)
+        out = pd.concat(
+            [
+                left_join_part,
+                right_join_part,
+            ],
+            axis=1,
+        )
+        out.set_index("_hipscat_index", inplace=True)
+        out["_DIST"] = distances
+
+        return out
+
+    def _find_crossmatch_indices(self, n_neighbors):
+        # calculate the gnomic distances to use with the KDTree
+        clon, clat = hp.pix2ang(
+            hp.order2nside(self.left_order), self.left_pixel, nest=True, lonlat=True
+        )
+        xy1 = self._frame_gnomonic(
+            self.left, self.left_metadata.catalog_info, clon, clat
+        )
+        xy2 = self._frame_gnomonic(
+            self.right, self.right_metadata.catalog_info, clon, clat
```

Comment:
This is using the lon/lat from the left side pixel distance only. That's ok?

Reply:
As far as I understand, they need to be in the same frame, centered around the same point. And since we're crossmatching them, the left pixel center should be close enough to both that it still works fine as an approximation. But this is Mario/Sam's algorithm, so asking them for clarification might be good